policyeval/broker.go

```diff
@@ -0,0 +1,347 @@
+	"github.com/hashicorp/nomad-autoscaler/helper/uuid"
+	"github.com/hashicorp/nomad-autoscaler/sdk"
+)
+
+// Broker stores, dedups and control access to policy evaluation requests.
+type Broker struct {
+	logger hclog.Logger
+
+	l sync.RWMutex
+
+	// nackTimeout is the time required for a dequeued eval to be ack'd.
+	nackTimeout time.Duration
+
+	// deliveryLimit is the number of times an eval can be dequeued before
+	// being considered as failed.
+	deliveryLimit int
+
+	// pendingEvals holds evaluations that are ready to be picked for
+	// further evaluation.
+	// Each key represents a different queue and the value is a heap holding
+	// the evals in order of priority and create time.
+	pendingEvals map[string]PendingEvaluations
+
+	// enqueuedEvals and enqueuedPolicies are indexes to quickly find if an eval
+	// or a policy is waiting be evaluated.
+	enqueuedEvals    map[string]int
+	enqueuedPolicies map[string]string
+
+	// unack tracks evaluations that have not been ack'd yet.
+	unack map[string]*unackEval
+
+	// waiting tracks Dequeue requests that are blocked waiting for work.
+	waiting map[string]chan struct{}
+}
+
+// unackEval tracks an unacknowledged evaluation along with the Nack timer
+type unackEval struct {
+	Eval      *sdk.ScalingEvaluation
+	Token     string
+	NackTimer *time.Timer
+}
+
+// NewBroker returns a new Broker object.
+func NewBroker(l hclog.Logger, timeout time.Duration, deliveryLimit int) *Broker {
+	return &Broker{
+		logger:           l.Named("broker"),
+		nackTimeout:      timeout,
+		deliveryLimit:    deliveryLimit,
+		pendingEvals:     make(map[string]PendingEvaluations),
+		enqueuedEvals:    make(map[string]int),
+		enqueuedPolicies: make(map[string]string),
+		unack:            make(map[string]*unackEval),
+		waiting:          make(map[string]chan struct{}),
+	}
+}
+
+// Enqueue adds an eval to the broker.
+func (b *Broker) Enqueue(eval *sdk.ScalingEvaluation) {
+	b.l.Lock()
+	defer b.l.Unlock()
+	b.enqueueLocked(eval, "")
+}
+
+func (b *Broker) enqueueLocked(eval *sdk.ScalingEvaluation, token string) {
+	logger := b.logger.With(
+		"eval_id", eval.ID, "policy_id", eval.Policy.ID,
+		"queue", eval.Policy.Type, "token", token)
+
+	logger.Debug("enqueue eval")
+
+	// Check if eval is already enqueued.
+	if _, ok := b.enqueuedEvals[eval.ID]; ok {
+		if token == "" {
+			logger.Debug("eval already enqueued")
+			return
+		}
+	} else {
+		b.enqueuedEvals[eval.ID] = 0
+	}
+
+	queue := eval.Policy.Type
+
+	// Get pending heap for the policy type.
+	pending, ok := b.pendingEvals[queue]
+	if !ok {
+		// Initialize pending heap and waiting channel.
+		pending = PendingEvaluations{}
+		if _, ok := b.waiting[queue]; !ok {
+			b.waiting[queue] = make(chan struct{}, 1)
+		}
+	}
+
+	// Check if an eval for the same policy is already enqueued.
+	pendingEvalID, ok := b.enqueuedPolicies[eval.Policy.ID]
+	if !ok {
+		b.enqueuedPolicies[eval.Policy.ID] = eval.ID
+	} else if pendingEvalID != eval.ID {
```

Comment:
If `pendingEvalID == eval.ID` it should actually be caught by the block above:

```go
if _, ok := b.enqueuedEvals[eval.ID]; ok {
	if token == "" {
		logger.Debug("eval already enqueued")
		return
	}
}
```

The cases for submitting the same the same ID twice (that I can think of) are:
* eval got NACK'd
* the policy manager sends it to the broker twice (for some reason....)

In the first case, the `token` is not empty, so the flow will proceed until this line and go through since the eval ID will be the same. But since it will have same `CreateTime` it will return without doing anything (something like this [test case](https://github.com/hashicorp/nomad-autoscaler/pull/282/files#diff-c600eef77639493229b591b692f5d3d7R109-R115)).

The second case will exit early in the block I mentioned earlier since the token will be empty (there's a [test case for this](https://github.com/hashicorp/nomad-autoscaler/pull/282/files#diff-c600eef77639493229b591b692f5d3d7R89-R91)).